crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,654 @@
+}
+
+#[test]
+fn reject_expired_at_too_long() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let created_at = crate::now();
+    let expires_at =
+        created_at + std::time::Duration::from_secs(60 * 40).as_micros() as i64;
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidExpiresAt"));
+}
+
+#[test]
+fn reject_bad_sig() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        signature: Some("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidSignature"));
+}
+
+#[test]
+fn default_storage_rollover() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let addr = s.listen_addr();
+    let mut test_prop: u32 = 0;
+    let mut put_info = move || {
+        use base64::prelude::*;
+        let mut agent_seed = [0; 32];
+        agent_seed[..4].copy_from_slice(&test_prop.to_le_bytes());
+        let agent_seed = BASE64_URL_SAFE_NO_PAD.encode(agent_seed);
+        PutInfo {
+            addr,
+            agent_seed: &agent_seed,
+            test_prop: &format!("{test_prop}"),
+            ..Default::default()
+        }
+        .call()
+        .unwrap();
+        test_prop += 1;
+    };
+
+    let addr = s.listen_addr();
+    let get = move || {
+        let addr = format!("http://{:?}/bootstrap/{}", addr, S1);
+        let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+        let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+        res.into_iter().map(|m| m.test_prop).collect::<Vec<_>>()
+    };
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    assert_eq!(
+        res.as_slice(),
+        &[
+            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
+            "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23",
+            "24", "25", "26", "27", "28", "29", "30", "31",
+        ]
+    );
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    assert_eq!(
+        res.as_slice(),
+        &[
+            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
+            "13", "14", "15", "48", "49", "50", "51", "52", "53", "54", "55",
+            "56", "57", "58", "59", "60", "61", "62", "63",
+        ]
+    );
+}
+
+#[test]
+fn multi_thread_stress() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = s.listen_addr();
+
+    let start = std::time::Instant::now();
+
+    // the testing config has 2 worker threads.
+    // Read and write with more than that.
+    const T_W_COUNT: u32 = 16;
+    const T_R_COUNT: u32 = 32;
```

Comment:
I'm curious about what we would expect in production environments.

On my machine when I bump the T_R_COUNT above 300 & start getting this error:

```
thread '<unnamed>' panicked at crates/bootstrap_srv/src/test.rs:561:18:
called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: Transport { kind: Io, message: None, url: Some(Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(37587), path: "/bootstrap/2o79pTXHaK1FTPZeBiJo2lCgXW_P0ULjX_5Div_2qxU/ovovSjVboukHpTAJ6eN8rd96x-ZqCLoHYx9VMHKz8kw", query: None, fragment: None }), source: Some(Custom { kind: ConnectionReset, error: Transport(Transport { kind: Io, message: Some("Error encountered in the status line"), url: None, source: Some(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }) }) }) } }
```

Reply:
This will also be mitigated somewhat when we implement rate limiting, and if we front the server with cloudflare.